srcts/src/components/_shinyRemovedObserver.ts

```diff
@@ -0,0 +1,106 @@
+type Callback<T> = (el: T) => void;
+
+/**
+ * Watch for the removal of specific elements from regions of the page.
+ */
+export class ShinyRemovedObserver {
+  private observer: MutationObserver;
+  private watching: Set<HTMLElement>;
+
+  /**
+   * Creates a new instance of the `ShinyRemovedObserver` class to watch for the
+   * removal of specific elements from part of the DOM.
+   *
+   * @param selector A CSS selector to identify elements to watch for removal.
+   * @param callback The function to be called on a matching element when it
+   * is removed.
+   */
+  constructor(selector: string, callback: Callback<HTMLElement>) {
+    this.watching = new Set<HTMLElement>();
+    this.observer = new MutationObserver((mutations) => {
+      const found = [];
+      for (const { type, removedNodes } of mutations) {
+        if (type !== "childList") continue;
+        if (removedNodes.length === 0) continue;
+
+        for (const node of removedNodes) {
+          if (!(node instanceof HTMLElement)) continue;
+          if (node.matches(selector)) {
+            found.push(node);
+          }
+          if (node.querySelector(selector)) {
+            node
+              .querySelectorAll<HTMLElement>(selector)
+              .forEach((el) => found.push(el));
+          }
+        }
+      }
+      if (found.length === 0) return;
+      found.forEach(callback);
+    });
+  }
+
+  /**
+   * Starts observing the specified element for removal of its children. If the
+   * element is already being observed, no change is made to the mutation
+   * observer.
+   * @param el The element to observe.
+   */
+  observe(el: HTMLElement): void {
+    const changed = this._flush();
+    if (this.watching.has(el)) {
+      if (!changed) return;
+    } else {
+      this.watching.add(el);
+    }
+
+    if (changed) {
+      this._restartObserver();
+    } else {
+      this.observer.observe(el, { childList: true, subtree: true });
+    }
+  }
+
+  /**
+   * Stops observing the specified element for removal.
+   * @param el The element to unobserve.
+   */
+  unobserve(el: HTMLElement): void {
+    if (!this.watching.has(el)) return;
+    // MutationObserver doesn't have an "unobserve" method, so we have to
+    // disconnect and re-observe all elements that are still being watched.
+    this.watching.delete(el);
+    this._flush();
+    this._restartObserver();
+  }
```

Comment:
I think I'd prefer this to be commented out with a comment about how it currently isn't being used

Reply:
I'd rather not, I guess because blocks of commented code give me the ick. I'd prefer to delete than to comment out code, but I'd also prefer to have a complete implementation than a partial one. Maybe we run into a similar problem and can pick up this class and drop it in. Or maybe we delete this whole file in a few weeks.